### ISSUE TITLE
verifier: map ptr bug in tail call

### ISSUE THREAD

**Issue by borkmann** [cilium/cilium#1033]
r2 (map pointer) should never be 0 here:
```
2186: (15) if r2 == 0x0 goto pc+9
 R0=inv R1=inv63,min_align=1 R2=inv R3=imm2,min_value=2,max_value=2,min_align=2 R4=inv58,min_value=0,max_value=60,min_align=4 R5=ctx R6=ctx R7=pkt(id=0,off=0,r=34) R8=inv R9=inv57,min_value=14,max_value=74,min_align=2 R10=fp
2187: (b7) r1 = 2
2188: (63) *(u32 *)(r5 +48) = r1
2189: (63) *(u32 *)(r5 +52) = r8
2190: (bf) r1 = r5
2191: (18) r2 = 0x0
2193: (b7) r3 = 2
2194: (85) call bpf_tail_call#12
R2 type=imm expected=map_ptr

Error filling program arrays!
```

**Comment by borkmann** [comment]
No issue with following code:
```
static inline struct endpoint_info *__inline__ lookup_ip4_endpoint(struct iphdr *ip4)
{
        struct endpoint_key key = { };
        key.ip4 = bpf_ntohl(42);
        return map_lookup_elem(&cilium_lxc, &key);
}
```
Issue with that one:
```
static inline struct endpoint_info *__inline__ lookup_ip4_endpoint(struct iphdr *ip4)
{
        struct endpoint_key key = {
                .ip4 = bpf_ntohl(42);
        };

        return map_lookup_elem(&cilium_lxc, &key);
}
```
asm diff:
```
# diff ./bpf_lxc.S bpf_lxc.S.issue
2303,2304c2303,2304
< 	ldw	r8, 30(r6)
< 	mov	r1, r8
---
> 	ldw	r7, 30(r6)
> 	mov	r1, r7
2309,2313c2309,2315
< 	mov	r7, 0
< 	std	-56(r10), r7
< 	std	-64(r10), r7
< 	mov	r1, 704643072
< 	stw	-64(r10), r1
---
> 	ld_64	r1, <MCOperand Expr:(.Llookup_ip4_endpoint.key)>
> 	ldw	r2, 0(r1)
> 	ldw	r1, 4(r1)
> 	slli	r1, 32
> 	or	r1, r2
> 	std	-56(r10), r1
> 	std	-64(r10), r1
2374,2377c2376,2380
< 	std	-120(r10), r7
< 	std	-128(r10), r7
< 	andi	r8, 65535
< 	stw	-128(r10), r8
---
> 	mov	r6, 0
> 	std	-120(r10), r6
> 	std	-128(r10), r6
> 	andi	r7, 65535
> 	stw	-128(r10), r7
2380,2383c2383,2386
< 	stw	-40(r10), r7
< 	std	-48(r10), r7
< 	std	-56(r10), r7
< 	std	-64(r10), r7
---
> 	stw	-40(r10), r6
> 	std	-48(r10), r6
> 	std	-56(r10), r6
> 	std	-64(r10), r6
2397d2399
< 	mov	r6, 0
2414c2416,2417
< 	jsgt	r6, r0 goto LBB5_309
---
> 	mov	r1, 0
> 	jsgt	r1, r0 goto LBB5_309
8535a8539,8544
> 
> 	.section	.rodata.cst16,"aM",@progbits,16
> 	.align	4                       # @lookup_ip4_endpoint.key
> .Llookup_ip4_endpoint.key:
> 	.long	704643072               # 0x2a000000
> 	.zero	12
```

**Comment by borkmann** [comment]
LLVM generates an invalid reloc entry.

tc says:
```
ELF contains non-map related relo data in entry 11 pointing to section 23! Compiler bug?!
```
And will then stop populating remaing relo entries, hence the imm 0 instead of map pointer at a different location.

More details from asm:
```
[...]
# BB#154:
        std     -136(r10), r5
        ld_64   r1, <MCOperand Expr:(.Llookup_ip4_endpoint.key)>
        ldw     r2, 0(r1)
        ldw     r1, 4(r1)
        slli    r1, 32
        or      r1, r2
        std     -56(r10), r1
        std     -64(r10), r1
        mov     r2, r10
        addi    r2, -64
        ld_64   r1, <MCOperand Expr:(cilium_lxc)>
        call    1
        jeqi    r0, 0 goto LBB5_305
[...]
        .section        .rodata.cst16,"aM",@progbits,16
        .align  4                       # @lookup_ip4_endpoint.key
.Llookup_ip4_endpoint.key:
        .long   704643072               # 0x2a000000
        .zero   12
```

**Comment by borkmann** [comment]
Opened LLVM ticket: https://bugs.llvm.org/show_bug.cgi?id=33599

**Comment by borkmann** [comment]
iproute2: http://patchwork.ozlabs.org/patch/780937/

**Comment by borkmann** [comment]
Fixed by Yonghong in LLVM 5.0.0 trunk.
